### PR TITLE
fix #3294 #3303 exposing apigroups and apiresource

### LINKED ISSUE #3294: how to get apiGroupList use kubernetes-client?
how to get apiGroupList use kubernetes-client?

### PR DESCRIPTION
**Pull request by shawkins**
## Description
Adds client methods for apigroups and apiresources.  Also consolidates existing logic about openshift adapting.

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

**Comment by shawkins** [comment]
This adds methods to the Client interface for the api resource and group information.  Alternatives include introducing a Client.metadata() call and associated container api, but something like that hasn't been mentioned anywhere so it doesn't seem necessary.

**Comment by shawkins** [comment]
I'll address the openshift test next time I have a cluster up to test against.  The compile errors though don't make sense to me.

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/Client.java [hunk 1]
```diff
@@ -48,5 +50,19 @@
...
   boolean supportsApiPath(String path);
 
+  @Override
   void close();
+  
+  /**
+   * Returns the api groups
+   * @return the {@link APIGroupList} metadata
+   */
+  APIGroupList getApiGroups();
```
Comment (rohanKanojia) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/Client.java:60]: Do you think we should also add another method for fetching a single APIGroup with name?:
```java
APIGroup getApiGroup(String name)
```

**Comment by rohanKanojia** [comment]
> I'll address the openshift test next time I have a cluster up to test against. 

@shawkins : I'm not sure you're aware of this but you can easily get a free OpenShift Cluster with Red Hat Developer Sandbox[0]. I usually test OpenShift related stuff and it works okay.

> The compile errors though don't make sense to me.

I'm not able to reproduce these errors as well.

[0] https://developers.redhat.com/developer-sandbox

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/Client.java (hunk 1 above)
Comment (shawkins) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/Client.java:60]: @rohanKanojia I'm fine with adding that.

**Comment by shawkins** [comment]
Looks like I made a simple mistake in the isAdaptable method.  I've pushed a correction along with getting a single api group.

**Comment by shawkins** [comment]
@rohanKanojia the remaining exception is related to https://github.com/fabric8io/kubernetes-client/pull/3405/files#diff-32314261b10bfe7703677b8086d5a28ff2dd1d0a8dc1061cdaad58e22c091576L84 I've changed things to no longer swallow the exception - the mock server is not responding to /apis.  Was the old code intentionally swallowing the exception or should I look at adding mock server support for this as well?

**Comment by rohanKanojia** [comment]
I think I did a mistake by swallowing the exception.

**Comment by shawkins** [comment]
@rohanKanojia I think things are good now.  There was another boolean that needed flipped.  The build still seems to be hitting those weird compile errors.

**Comment by shawkins** [comment]
@manusa rebased and still behaving oddly.

**Comment by manusa** [comment]
Fixed in #3487

**Pull request closed without merging**